### PR TITLE
[New][1/N] Dynamo skipfiles refactor

### PR DESCRIPTION
**Pull request by yanboliang**
This is the replacement of #109567. Now I preserved all existing semantics and only focusing on API (for developers) and code structure changes. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Comment by yanboliang** [comment]
@ezyang I'm sure I preserved all existing semantics now. If not, I owe you a beer, lol.

**Review by ezyang — APPROVED** [review]
I did not carefully test equivalence. It is not so easy to test the function for equivalence but there are data structures it references which I think you should manually verify.

Please check comments

File: torch/_dynamo/config.py [hunk 1]
```diff
@@ -144,21 +144,8 @@
...
-    torch.ao.nn,
-    torch._refs,
-    torch._prims,
-    torch._decomp,
-    torch.utils._contextlib,
-    torch.utils._pytree,
-    torch.fx._pytree,
-    torch.sparse,
-}
+# legacy config, does nothing now!
```
[the remark this answers is not in the document]
Comment (yanboliang) [reply torch/_dynamo/config.py:147]: Searched in fbcode, nobody is using it, so it's safe for this change.

**Comment by voznesenskym** [comment]
This is breaking our DDP tracing branch, can we revert? @fegin

**Comment by voznesenskym** [comment]
I don't care if we revert or forward fix. cc @ezyang @yanboliang 

Target PR is here https://github.com/pytorch/pytorch/pull/110662/files

I'll let yall decide, and I can fix whichever way we decide.

**Comment by yanboliang** [comment]
@voznesenskym We preserve all previous semantics in this PR, just do code refactor w/o any change on semantics. Can you tell me which test is failed because of this PR? I looked at these failures in #110662, I didn't see which one caused by this. If there is, it would be super helpful for us to hardening these code. I'm pretty happy to debug together if you have time.

**Comment by fegin** [comment]
@yanboliang The issue is importing `_functional_collectives` too soon. I don't know why this affects the tracing. But you can use https://github.com/pytorch/pytorch/pull/110688 to reproduce.  https://github.com/pytorch/pytorch/pull/110688 adds a hack to delay the import of `_functional_collectives`, which should be removed once we figure out the root cause.

This also means that all the functional collective cannot be compiled cc., @wanchaol

**Comment by voznesenskym** [comment]
This is an important PR, lets forward fix, not revert. 

It sounds like a simple ordering bug, not worth the jitter of a revert.

**Comment by yanboliang** [comment]
@fegin Thanks for your repro, I almost know the reason, another import order caused problem. I'll look at your repro and confirm if my guess is correct. Meanwhile, I'm working on another PR which will remove these import order caused issue permanently. 
@voznesenskym Thanks for your understanding!

**Comment by yanboliang** [comment]
Confirmed this is import order caused problem, I hope my next PR can fundamentally fix the series of such problem.

**Comment by wanchaol** [comment]
> Confirmed this is import order caused problem, I hope my next PR can fundamentally fix the series of such problem.

Maybe as a follow up we need to figure out why this issue wasn't caught in CI, we have bunch of torch.compile + functional collective tests, but maybe they are somehow skipped by CI https://github.com/pytorch/pytorch/blob/main/test/distributed/test_inductor_collectives.py#L572

**Pull request closed without merging**